### PR TITLE
Add B helpers to check if the Sv is a boolean

### PR DESCRIPTION
**Pull request by atoomic**
Bool are using PVNV. It makes it more convenient
to provide these helpers at the top level for any SVs.

So we can easily check if the SV is a boolean and check
if it's true or false.

Note: alternatively we could consider a single helper to return the boolean value:
- 1 if bool & Yes
- 0 if bool & No
- undef if not bool

### CODE REVIEW

**Review by toddr — APPROVED** [review]

**Comment by leonerd** [comment]
Not sure I get this. The `builtin::` namespace already provides an `is_bool` function that can be called from pure perl. There's no need to add it this to `B`. Or have I somehow missed the point here?

**Comment by atoomic** [comment]
B serves as an authoritative interface for the internal and needs to be up to date to parse these new additions.
It should be in blead before releasing 5.36.0

**Comment by toddr** [comment]
> Not sure I get this. The `builtin::` namespace already provides an `is_bool` function that can be called from pure perl. There's no need to add it this to `B`. Or have I somehow missed the point here?

The builtin functions against an SV. It does not help though with an arbitrary B::SV when walking a tree.

**Comment by demerphq** [comment]
On Tue, 12 Apr 2022, 04:22 ℕicolas ℝ., ***@***.***> wrote:

> B serves as an authoritative interface for the internal and needs to be up
> to date to parse these new additions.
> It should be in blead before releasing 5.36.
>

+1

Yves

**Comment by Leont** [comment]
The camel-case seems inconsistent with the rest of the API

**Comment by toddr** [comment]
> The camel-case seems inconsistent with the rest of the API

If there is a consistent naming scheme, I'd say we go with something like `is_offset` ?

**Comment by atoomic** [comment]
Thanks for your feedbacks, I applied a few suggestions:
* use SvTRUE
* rename functions to `is_bool`, `is_bool_yes` and `is_bool_no`

**Comment by rjbs** [comment]
@leonerd Can this go in before v5.35.11?  I don't see why not?

File: ext/B/B.pm
Comment (leonerd) [inline ext/B/B.pm]: "... is a boolean and true". Most regular SVs that are boolean and true will not be exactly `PL_sv_yes`

File: ext/B/B.xs
Comment (leonerd) [inline ext/B/B.xs]: Why the ternary? This can simply be

```
RETVAL = SvIsBOOL(sv);
```

Ditto the other two functions below.

File: ext/B/B.pm
Comment (tonycoz) [inline ext/B/B.pm]: The code doesn't check whether the SV is PL_Yes or PL_No, it checks whether the SV is a boolean (which any SV can be) and for the second two, whether that value is true or false.

File: ext/B/B.xs
Comment (leonerd) [reply ext/B/B.xs]: Actually for that matter, why does it have its own name? Why can it not simply be the SvIsBOOL itself?

```
bool
SvIsBOOL(sv)
    B::SV  sv
```

with no code body at all. E.g. compare to the `SvIV` below.

File: ext/B/B.xs
Comment (leonerd) [inline ext/B/B.xs]: I'm not sure why this is useful (also the `no` below it). If anyone cares to check they can test it from perl-land just as easily

```perl
  if($b_sv->IsBOOL && $b_sv->TRUE) { ... }
```

Comment (toddr) [reply ext/B/B.xs]: When you are serializing the information, you need a way to query the B object so it round trips properly in another process.

Comment (atoomic) [reply ext/B/B.xs]: I think this provides a better abstraction, where you could change internal without changing all consumers
if this would change.

File: ext/B/B.xs
Comment (atoomic) [reply ext/B/B.xs]: that could work

File: ext/B/B.pm
Comment (atoomic) [reply ext/B/B.pm]: going to update it as suggested

File: ext/B/B.pm
Comment (atoomic) [reply ext/B/B.pm]: going to adjust the description to match the suggestion for is_bool_yes

File: ext/B/B.xs
Comment (leonerd) [reply ext/B/B.xs]: Yes. But why do you need to add a new XS method on the B::SV class (`is_bool_yes`) which just returns information that is the `and` combination of two other methods that already exist (`IsBOOL and TRUE`)?

Comment (leonerd) [reply ext/B/B.xs]: > I think this provides a better abstraction, where you could change internal without changing all consumers
if this would change.

This isn't about internals. IsBOOL returns true for any boolean. TRUE returns true for any value that perl considers true. Thus the combination `$sv->IsBOOL and $sv->TRUE` is simply the combination of those two facts. It doesn't need a special method.

This would be like adding a special method to ask if a given SV is numerical and the value 1. Why bother? We can just check those two facts, separately, using the regular query methods. No need to special-case it.

Comment (atoomic) [reply ext/B/B.xs]: This is assuming that this will always be the way to check if the boolean is true or false.

I was under the assumption that this could change overtime and we could start using one shared SV for true or false at one point or whatever other optimization we decide in the future.

If we think that `IsBOOL and TRUE` is always going to work, then indeed we can consider removing the `is_bool_yes` and `is_bool_no` helpers. I was trying to provide abstraction for the user which should not track how this is implemented behind.

This would add more boilerplate to code parsing a B tree and trying to check booleans. But I'm fine with that.

File: ext/B/B.xs
Comment (tonycoz) [inline ext/B/B.xs]: Aside from @leonerd 's comment (which I agree with), this will invoke magic if any on the SV if it started out as a bool.

Other B accessors correctly don't do that, since the intent is to fetch the internal state of the SV.

This (or a replacement TRUE method if the combos are removed) should use SvTRUE_nomg() rather than SvTRUE().

File: ext/B/B.xs
Comment (atoomic) [reply ext/B/B.xs]: note that this should not use `SvIsBool` as there is no prefix for the `B::SV` package set
so I gave a try of using 
```
bool
IsBOOL(sv)
    B::SV   sv
```

this is now failing with
```
./perl: symbol lookup error: ../lib/auto/B/B.so: undefined symbol: IsBOOL
```

note sure how to fix it. If you have any ideas or suggestion I'm glad to take it
thanks

File: ext/B/B.xs
Comment (leonerd) [reply ext/B/B.xs]: I was imagining that there's already a `->TRUE` method which is a wrapper of `SvTRUE()` (though, as @tonycoz points out it really should be using `SvTRUE_mg()`. If not, then I'd much prefer adding that one as it's more universally useful - it would apply to any kind of SV then.

But yes, have no fear of `SvTRUE()` and friends ever changing - they are as much a core part of the basic SV API as something like `SvIV` is.

File: ext/B/B.xs
Comment (atoomic) [reply ext/B/B.xs]: ok need to setup the prefix... but then this become a mess
```diff
diff --git a/ext/B/B.xs b/ext/B/B.xs
index 878fb0dba5..4f75bec4f0 100644
--- a/ext/B/B.xs
+++ b/ext/B/B.xs
@@ -1398,12 +1398,12 @@ aux_list(o, cv)



-MODULE = B     PACKAGE = B::SV
+MODULE = B     PACKAGE = B::SV         PREFIX = Sv

 #define MAGICAL_FLAG_BITS (SVs_GMG|SVs_SMG|SVs_RMG)

 U32
-REFCNT(sv)
+SvREFCNT(sv)
        B::SV   sv
     ALIAS:
        FLAGS = 0xFFFFFFFF
@@ -1417,7 +1417,7 @@ REFCNT(sv)
        RETVAL

 void
-object_2svref(sv)
+Svobject_2svref(sv)
        B::SV   sv
     PPCODE:
        ST(0) = sv_2mortal(newRV(sv));
@@ -1428,7 +1428,7 @@ SvIsBOOL(sv)
     B::SV   sv

 bool
-is_bool_yes(sv)
+Svis_bool_yes(sv)
     B::SV   sv
     CODE:
     RETVAL = SvIsBOOL(sv) && SvTRUE(sv);
@@ -1436,7 +1436,7 @@ is_bool_yes(sv)
         RETVAL

 bool
-is_bool_no(sv)
+Svis_bool_no(sv)
     B::SV   sv
     CODE:
     RETVAL = SvIsBOOL(sv) && !SvTRUE(sv);
```

Comment (atoomic) [reply ext/B/B.xs]: Would you really prefer this flavor?

File: ext/B/B.xs
Comment (atoomic) [reply ext/B/B.xs]: yeap I can use SvTRUE_nomg... 
my first implementation was using a PVX check as the `Perl_sv_isbool` does.
But as suggested I changed it to `SvTRUE` and you are right `SvTRUE_nomg` is a better choice
I wonder if the `PVX` check was not the correct way to do it
```
 PERL_STATIC_INLINE bool
 Perl_sv_isbool(pTHX_ const SV *sv)
 {
     return SvIOK(sv) && SvPOK(sv) && SvIsCOW_static(sv) &&
         (SvPVX_const(sv) == PL_Yes || SvPVX_const(sv) == PL_No);
 }
```

File: ext/B/B.xs
Comment (Leont) [reply ext/B/B.xs]: Using PREFIX generally only makes sense if you don't have a CODE body

File: ext/B/B.xs
Comment (leonerd) [reply ext/B/B.xs]: The `SvPVX_const` test is how the implementation of SvIsBOOL itself is implemented (along with the IsCOW_static check). That function is the one that implements the macro that everyone *else* should be using.

**Comment by atoomic** [comment]
Hi here is another iteration via 24204f27c73c1682d8a51e549369f96a3e621c5e trying to use most suggestions from this PR:
Note that it has the following issues:
- add the Sv prefix to B::SV to use a no body helper for IsBOOL
- IsBOOL is now back to camel case after requesting a no body implementation
- back to original PVX check rather than SvTRUE to avoid triggering magic

Question:
if we decide to remove the `is_bool_yes` and `is_bool_no` helpers, would not this trigger some magic
while trying to call `$sv->TRUE` on it?

**Comment by leonerd** [comment]
> if we decide to remove the is_bool_yes and is_bool_no helpers, would not this trigger some magic
while trying to call $sv->TRUE on it?

That's why you'd have `$sv->TRUE` call the SvTRUE_mg macro:

```
bool
SvTRUE(sv)
    B::SV sv
    CODE:
      RETVAL = SvTRUE_nomg(sv);
    OUTPUT:
      RETVAL
```

Having done that, any caller who cares can check `$sv->IsBOOL and $sv->TRUE` so you no longer need your (still incorrectly implemented) is_bool_yes / no.

**Comment by toddr** [comment]
> so you no longer need your (still incorrectly implemented) is_bool_yes / no.

Point of clarification. Are you suggesting that this existing code is incorrectly implemented?

```C
PERL_STATIC_INLINE bool
Perl_sv_isbool(pTHX_ const SV *sv)
{
    return SvIOK(sv) && SvPOK(sv) && SvIsCOW_static(sv) &&
        (SvPVX_const(sv) == PL_Yes || SvPVX_const(sv) == PL_No);
}
```

**Comment by leonerd** [comment]
No. Please read again my comment:

> The SvPVX_const test is how the implementation of SvIsBOOL itself is implemented (along with the IsCOW_static check). That function is the one that implements the macro that everyone else should be using.

The code you quoted is the *implementation* of SvIsBOOL. It is permitted to use its own internal trickery. Everyone else should call that macro instead, so as to be abstracted away from its implementation.

**Comment by leonerd** [comment]
To clarify further: `SvIsBOOL`'s current implementation relies on the internals of how real-boolean SVs work - they are COW-static SVs, whose PV pointer is exactly a pointer to the `PL_Yes` or `PL_No` static string buffers. If we have a COW-static SV whose PV pointer is one of these, then we know we have a kind of boolean.

Any other user code, knowing *that* we have a real-boolean (by calling the above macro) does not need to know anything about these internals. The only thing that distinguishes it, beyond this, is whether the boolean is true or false. For that the user can call the plain ol' `SvTRUE()` or `SvTRUE_nomg()` macros to test for SV truth exactly the same way as they've always done for the past 20 years.

I will say once more in case it is still unclear: There is **absolutely no need** for any code *except the implementation of `SvIsBOOL`* to be making any sort of pointer comparison against the `PL_Yes` or `PL_No` static arrays.

File: ext/B/B.xs
Comment (haarg) [reply ext/B/B.xs]: I'm not particularly a fan of `is_bool_yes` or `is_bool_no`, but more importantly:

B is not the place for abstractions. If an abstract understanding was appropriate, it would be available through normal perl ops. When the internals change, B should reflect that. If you were relying on those internals, your code should break.

**Review by leonerd — CHANGES_REQUESTED** [review]
The `IsBOOL` part looks fine now.

The `is_bool_yes` would be better done by making a new `TRUE` in similar style to `IsBOOL` and letting the user test `$sv->IsBOOL and $sv->TRUE` rather than having this special-case method just for that test. Similar thoughts for `is_bool_no`.

As commented; make sure to call `SvTRUE_nomg()` to avoid the magic effects of the regular `SvTRUE` macro.

**Comment by atoomic** [comment]
I think following @toddr advice a better implementation would be to use a single helper `check_bool` or whatever name
that could return three different values:
- undef: when it's not a boolean 
- PL_sv_yes: when it s a boolean and true
- PL_sv_no: when it s a boolean and false 

So then this would be easy to check, and avoid multiple calls and helpers
```perl
my $bool_state = $sv->check_bool;
if ( defined $bool_state ) {
   say $bool_state == 1 ? 'true' : 'false';
} else {
... not a boolean
}
```

**Comment by tonycoz** [comment]
> I think following @toddr advice a better implementation would be to use a single helper `check_bool` or whatever name that could return three different values:
> 
>     * undef: when it's not a boolean
> 
>     * PL_sv_yes: when it s a boolean and true
> 
>     * PL_sv_no: when it s a boolean and false

I think you're trying to over-optimize to the point of providing a bad UI.

It's a UI perl itself uses for mutators, eg read, readline, but in those examples the underlying state changes due to the call itself, which isn't the case here.

Ideally I think the methods should do one thing and not try to overload their return values in this way.

**Comment by atoomic** [comment]
I gave it a try via ad594907a94cbdd20adf6c4d0448f1caed2af174 and IMO this looks much cleaner and less noise added to the B::SV package. One single helper instead of three...
Definitively open to any opinion/feedbacks...

**Comment by iabyn** [comment]
I dislike this intensely. It is mixing two very different concepts - to  retrieve the type of a variable and retrieve the value of a variable.

**Comment by leonerd** [comment]
I agree with @iabyn and @tonycoz here.

The B interface is supposed to be about as low-level and lightweight as possible; providing a perl interface layer onto basically the same "SV field accessor" macros that the C/XS layer already uses. C/XS already uses function-like macros called SvIsBOOL and SvTRUE to answer the questions you are trying to ask of it. Therefore the B interface layer really should provide the same. (That it previously didn't provide a `->TRUE` method already is perhaps an oversight of the current API).

The helper wrapper you provide *could* be useful in your use-case, but if you need it for your own use it would be easy to make as a one-line injected method in pureperl on top of the `->IsBOOL` and `->TRUE` accessors I've already suggested and given implementations for:

```perl
sub B::SV::check_bool ($sv) { return $sv->IsBOOL ? $sv->TRUE : undef }
```

File: ext/B/B.xs
Comment (toddr) [reply ext/B/B.xs]: > Aside from @leonerd 's comment (which I agree with), this will invoke magic if any on the SV if it started out as a bool.

@tonycoz: Is it possible to have a bool SV with magic on it?

**Comment by atoomic** [comment]
Thanks for your feedback, I've adjusted the PR by providing the `IsBOOL`, `TRUE` and `TRUE_nomg`.
TBH it's not clear to me once we have established that `IsBOOL` is true, if it's wrong to use `TRUE` instead of `TRUE_nomg`. Would not a boolean never contain magic? What would be the cases?

Waiting for another review.

**Comment by leonerd** [comment]
Any kind of SV can have magic on it. Don't forget that variables are SVs, and many kinds of magic apply to the variable, not the value. For example, a `tie`d lexical.

File: ext/B/B.pm
Comment (leonerd) [inline ext/B/B.pm:626]: Typo "whether"

**Review by leonerd — APPROVED** [review]
That's looking much better now. All the additions are simply wrappers for the same-named underlying C macros, as is the usual B style.

Approved, subject to one docs typo fix.

File: ext/B/B.pm
Comment (atoomic) [reply ext/B/B.pm:626]: thanks, fixed by e2f9e9f9

File: ext/B/B.pm
Comment (toddr) [inline ext/B/B.pm]: s/it's/it is/ ?

**Review by toddr — APPROVED** [review]

File: ext/B/B.pm
Comment (atoomic) [reply ext/B/B.pm]: fixed as part of c62e2787

**Comment by leonerd** [comment]
We've discussed this with PSC and decided the best thing to do with this is not to try to rush it in now, nor to extend the timetable to get a 5.35.12. Instead we'll wait for the next development cycle (5.37) to do so.

The new "defer-next-dev" label reflects this.

**Pull request merged**
